lib/spear/pool/pool_table.ex

```diff
@@ -0,0 +1,21 @@
+defmodule Spear.Pool.PoolTable do
+  use GenServer
+
+  def start_link(_opts) do
+    GenServer.start_link(__MODULE__, [])
+  end
+
+  def init([]) do
+    ets = :ets.new(__MODULE__, [:public, :named_table, {:read_concurrency, true}])
+    # We don't really care if we start at the 2nd connection, since it's simple round robin
+    counter_ref = :atomics.new(1, signed: false)
+    :atomics.put(counter_ref, 1, 0)
+    :ets.insert(__MODULE__, {:counter_ref, counter_ref})
+    {:ok, ets}
+  end
+
+  def read_and_increment do
+    counter_ref = :ets.lookup_element(__MODULE__, _key = :counter_ref, _pos = 2)
+    :atomics.add_get(counter_ref, 1, 1)
```

Comment:
It's hard to say without benchmarking it but I would suspect that it's faster to use `ets:update_counter/3` than to do two calls for `ets:lookup_element/3` and `atomics:add_get/3`.

Reply:
Thank you for the review,

> I'm hesitant to include pooling out-of-the-box because it's hard to know how users would want the routing to be balanced: I don't think there's a good one-size-fits-all solution.

I agree, I've removed the pooling implementation and added a pooling guide instead (since I believe it might still be useful for users).

> On the other hand, it might be faster even than `ets:update_counter/3` to eliminate the ets table here and use `persistent_term` to hold the atomic:
> 
> ```elixir
> def init([]) do
>   # setup counter_ref ...
>   :persistent_term.put(__MODULE__, counter_ref)
> end
> 
> def read_and_increment do
>   counter_ref = :persistent_term.get(__MODULE__)
>   :atomics.add_get(counter_ref, 1, 1)
> end
> ```
> 
> That could also eliminate the long-lived GenServer process since persistent_term doesn't need an owning process like ets does.

Good catch, I've integrated this change in the guide.

>  I think it should work to add configuration options to Spear.Connection for on_connect and on_disconnect which take anonymous functions (or maybe MFA tuples) that are executed when the connection connects and disconnects. What do you think?

That works perfectly, I've added these. The requirement is that the anonymous functions must be 0-arity functions (same with MFAs, and the arguments are ignored).

Let me know what you think, or how I could improve the PR.

Reply:
I've allowed to pass arguments to MFAs given to :on_connect (otherwise I could not figure out how to test them)  and I've added some tests.